File: CritterVision/app/src/main/java/com/example/crittervision/MainActivity.java
package com.example.crittervision;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.Preview;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;
import com.google.common.util.concurrent.ListenableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.drawable.ColorDrawable;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.view.View; // Required for View.LAYER_TYPE_HARDWARE (though not directly used in foreground method)
import android.widget.Button;
import android.widget.TextView;


public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";
    private VisionColorFilter.FilterType currentFilter = VisionColorFilter.FilterType.ORIGINAL;
    private static final int REQUEST_CODE_PERMISSIONS = 10;
    private static final String[] REQUIRED_PERMISSIONS = new String[]{android.Manifest.permission.CAMERA};
    private PreviewView previewView;
    private ExecutorService cameraExecutor;
    // private ImageView filteredImageView; // Decided against for now
    private TextView activeFilterTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        previewView = findViewById(R.id.previewView);
        activeFilterTextView = findViewById(R.id.activeFilterTextView);
        cameraExecutor = Executors.newSingleThreadExecutor();

        Button dogVisionButton = findViewById(R.id.dogVisionButton);
        Button catVisionButton = findViewById(R.id.catVisionButton);
        Button birdVisionButton = findViewById(R.id.birdVisionButton);
        Button originalVisionButton = findViewById(R.id.originalVisionButton);

        updateActiveFilterTextView(); // Set initial text

        dogVisionButton.setOnClickListener(v -> {
            currentFilter = VisionColorFilter.FilterType.DOG;
            updatePreviewFilter();
            updateActiveFilterTextView();
        });

        catVisionButton.setOnClickListener(v -> {
            currentFilter = VisionColorFilter.FilterType.CAT;
            updatePreviewFilter();
            updateActiveFilterTextView();
        });

        birdVisionButton.setOnClickListener(v -> {
            currentFilter = VisionColorFilter.FilterType.BIRD;
            updatePreviewFilter();
            updateActiveFilterTextView();
        });

        originalVisionButton.setOnClickListener(v -> {
            currentFilter = VisionColorFilter.FilterType.ORIGINAL;
            updatePreviewFilter();
            updateActiveFilterTextView();
        });

        if (allPermissionsGranted()) {
            startCamera();
        } else {
            ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS);
        }
    }

    private boolean allPermissionsGranted() {
        for (String permission : REQUIRED_PERMISSIONS) {
            if (ContextCompat.checkSelfPermission(this, permission) != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera();
            } else {
                Toast.makeText(this, "Permissions not granted by the user.", Toast.LENGTH_SHORT).show();
                // Consider disabling camera-dependent features or closing the app
            }
        }
    }

    private void startCamera() {
        ListenableFuture<ProcessCameraProvider> cameraProviderFuture = ProcessCameraProvider.getInstance(this);
        cameraProviderFuture.addListener(() -> {
            try {
                ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                Preview preview = new Preview.Builder().build();
                preview.setSurfaceProvider(previewView.getSurfaceProvider());

                CameraSelector cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA;

                cameraProvider.unbindAll(); // Unbind use cases before rebinding
                cameraProvider.bindToLifecycle(this, cameraSelector, preview);

            } catch (ExecutionException | InterruptedException e) {
                Log.e(TAG, "Use case binding failed", e);
                Toast.makeText(getApplicationContext(), "Error starting camera: " + e.getMessage(), Toast.LENGTH_LONG).show();
            }
        }, ContextCompat.getMainExecutor(this));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (cameraExecutor != null) {
            cameraExecutor.shutdown();
        }
    }

    private void updateActiveFilterTextView() {
        String filterName = "Original Vision"; // Default
        switch (currentFilter) {
            case DOG:
                filterName = "Dog Vision";
                break;
            case CAT:
                filterName = "Cat Vision";
                break;
            case BIRD:
                filterName = "Bird Vision";
                break;
            case ORIGINAL:
                // filterName is already "Original Vision"
                break;
        }
        activeFilterTextView.setText("Current View: " + filterName);
    }

    private void updatePreviewFilter() {
        ColorMatrixColorFilter filter = VisionColorFilter.getFilter(currentFilter);
        Log.d(TAG, "Updating filter to: " + currentFilter);

        Drawable foreground = previewView.getForeground();

        if (filter == null) {
            // Clear existing filter
            if (foreground != null) {
                // Performance Note: Applying ColorFilter to PreviewView's foreground can be inefficient for real-time camera streams.
                // If lag is observed, consider using CameraX Effects API (Preview.setEffect with a SurfaceProcessor from androidx.camera:camera-effects)
                // or an ImageAnalysis use case to process frames and display them on a separate ImageView for better performance.
                foreground.mutate().clearColorFilter();
                previewView.invalidate(); // Request redraw
                Log.d(TAG, "Cleared filter from PreviewView foreground.");
            } else {
                Log.d(TAG, "PreviewView foreground is null, no filter to clear.");
            }
        } else {
            // Apply new filter
            if (foreground == null && android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
                // Ensure foreground exists for API 23+ to apply a ColorFilter to it.
                // Using a transparent drawable as a base for the filter.
                previewView.setForeground(new ColorDrawable(Color.TRANSPARENT));
                foreground = previewView.getForeground();
                Log.d(TAG, "Initialized PreviewView foreground for filter application.");
            }

            if (foreground != null) {
                // Performance Note: Applying ColorFilter to PreviewView's foreground can be inefficient for real-time camera streams.
                // If lag is observed, consider using CameraX Effects API (Preview.setEffect with a SurfaceProcessor from androidx.camera:camera-effects)
                // or an ImageAnalysis use case to process frames and display them on a separate ImageView for better performance.
                foreground.mutate().setColorFilter(filter);
                previewView.invalidate(); // Request redraw
                Log.d(TAG, "Applied new filter to PreviewView foreground.");
            } else {
                Log.d(TAG, "PreviewView foreground is null, cannot apply filter.");
                 // Fallback or alternative methods could be logged or attempted here if foreground is not available/working.
                 // e.g., Log.w(TAG, "Consider using setEffect for PreviewView if foreground filtering is insufficient.");
            }
        }
    }
}

File: CritterVision/app/src/main/java/com/example/crittervision/MainActivity.kt
package com.example.crittervision

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.CameraSelector
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.common.util.concurrent.ListenableFuture
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class MainActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "MainActivity"
        private const val REQUEST_CODE_PERMISSIONS = 10
        private val REQUIRED_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)
    }

    private lateinit var previewView: PreviewView
    private lateinit var filterOverlay: View
    private lateinit var cameraExecutor: ExecutorService
    private lateinit var activeFilterTextView: TextView

    private var currentFilter: VisionColorFilter.FilterType = VisionColorFilter.FilterType.ORIGINAL

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        previewView = findViewById(R.id.previewView)
        activeFilterTextView = findViewById(R.id.activeFilterTextView)
        cameraExecutor = Executors.newSingleThreadExecutor()

        // Try to find an existing overlay view in the layout, or create one
        setupFilterOverlay()

        // Setup buttons
        val dogVisionButton: Button = findViewById(R.id.dogVisionButton)
        val catVisionButton: Button = findViewById(R.id.catVisionButton)
        val birdVisionButton: Button = findViewById(R.id.birdVisionButton)
        val originalVisionButton: Button = findViewById(R.id.originalVisionButton)

        updateActiveFilterTextView()

        dogVisionButton.setOnClickListener {
            currentFilter = VisionColorFilter.FilterType.DOG
            updatePreviewFilter()
            updateActiveFilterTextView()
        }
        catVisionButton.setOnClickListener {
            currentFilter = VisionColorFilter.FilterType.CAT
            updatePreviewFilter()
            updateActiveFilterTextView()
        }
        birdVisionButton.setOnClickListener {
            currentFilter = VisionColorFilter.FilterType.BIRD
            updatePreviewFilter()
            updateActiveFilterTextView()
        }
        originalVisionButton.setOnClickListener {
            currentFilter = VisionColorFilter.FilterType.ORIGINAL
            updatePreviewFilter()
            updateActiveFilterTextView()
        }

        if (allPermissionsGranted()) {
            startCamera()
        } else {
            ActivityCompat.requestPermissions(this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS)
        }
    }

    private fun setupFilterOverlay() {
        // Create a simple view that acts as our color overlay
        filterOverlay = View(this)
        filterOverlay.setBackgroundColor(Color.TRANSPARENT)
        filterOverlay.visibility = View.GONE

        // Add it to the content view as an overlay
        val contentView = findViewById<android.view.ViewGroup>(android.R.id.content)
        val layoutParams = android.view.ViewGroup.LayoutParams(
            android.view.ViewGroup.LayoutParams.MATCH_PARENT,
            android.view.ViewGroup.LayoutParams.MATCH_PARENT
        )
        contentView.addView(filterOverlay, layoutParams)

        Log.d(TAG, "Filter overlay created and added to content view")
    }

    private fun allPermissionsGranted(): Boolean {
        return REQUIRED_PERMISSIONS.all {
            ContextCompat.checkSelfPermission(baseContext, it) == PackageManager.PERMISSION_GRANTED
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                Toast.makeText(this, "Permissions not granted by the user.", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun startCamera() {
        val cameraProviderFuture: ListenableFuture<ProcessCameraProvider> = ProcessCameraProvider.getInstance(this)
        cameraProviderFuture.addListener({
            try {
                val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()

                val preview = Preview.Builder()
                    .build()

                preview.setSurfaceProvider(previewView.surfaceProvider)
                val cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA

                cameraProvider.unbindAll()
                cameraProvider.bindToLifecycle(this, cameraSelector, preview)

                Log.d(TAG, "Camera started successfully")

            } catch (e: Exception) {
                Log.e(TAG, "Use case binding failed", e)
                Toast.makeText(applicationContext, "Error starting camera: ${e.message}", Toast.LENGTH_LONG).show()
            }
        }, ContextCompat.getMainExecutor(this))
    }

    private fun updatePreviewFilter() {
        Log.d(TAG, "Applying filter: $currentFilter")

        when (currentFilter) {
            VisionColorFilter.FilterType.DOG -> {
                // Dog vision: Yellow overlay to simulate protanopia (red-green color blindness)
                filterOverlay.setBackgroundColor(Color.argb(120, 255, 255, 0)) // Semi-transparent yellow
                filterOverlay.visibility = View.VISIBLE
                Log.d(TAG, "Dog filter applied - yellow overlay visible")
                Toast.makeText(this, "🐕 Dog Vision Active: Yellow-dominant world", Toast.LENGTH_LONG).show()
            }
            VisionColorFilter.FilterType.CAT -> {
                // Cat vision: Blue-gray overlay for muted colors
                filterOverlay.setBackgroundColor(Color.argb(100, 100, 150, 200)) // Semi-transparent blue-gray
                filterOverlay.visibility = View.VISIBLE
                Log.d(TAG, "Cat filter applied - blue-gray overlay visible")
                Toast.makeText(this, "🐱 Cat Vision Active: Muted color perception", Toast.LENGTH_LONG).show()
            }
            VisionColorFilter.FilterType.BIRD -> {
                // Bird vision: Light purple overlay to simulate UV perception
                filterOverlay.setBackgroundColor(Color.argb(80, 200, 100, 255)) // Semi-transparent purple
                filterOverlay.visibility = View.VISIBLE
                Log.d(TAG, "Bird filter applied - purple overlay visible")
                Toast.makeText(this, "🦅 Bird Vision Active: UV-enhanced perception", Toast.LENGTH_LONG).show()
            }
            VisionColorFilter.FilterType.ORIGINAL -> {
                // Original vision: No overlay
                filterOverlay.visibility = View.GONE
                Log.d(TAG, "Original filter applied - overlay hidden")
                Toast.makeText(this, "👁️ Human Vision: Normal color perception", Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun updateActiveFilterTextView() {
        val filterName = when (currentFilter) {
            VisionColorFilter.FilterType.DOG -> "🐕 Dog Vision"
            VisionColorFilter.FilterType.CAT -> "🐱 Cat Vision"
            VisionColorFilter.FilterType.BIRD -> "🦅 Bird Vision"
            VisionColorFilter.FilterType.ORIGINAL -> "👁️ Human Vision"
        }
        activeFilterTextView.text = "Current View: $filterName"
        Log.d(TAG, "Updated filter text to: $filterName")
    }

    override fun onDestroy() {
        super.onDestroy()
        cameraExecutor.shutdown()
    }
}